### PR TITLE
[Admin][Order] Set customer IP on placed order

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | |
| License         | MIT |

zrzut ekranu 2016-11-17 o 11 54 48

### CODE REVIEW

File: src/Sylius/Behat/Page/Admin/Order/ShowPage.php [hunk 1]
```diff
@@ -392,6 +392,14 @@ public function hasBillingProvinceName($provinceName)
     /**
      * {@inheritdoc}
      */
+    public function hasIpAddressAssigner()
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Order/ShowPage.php]: `hasIpAddressAssigner` -> `hasIpAddressAssigned`

Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Order/ShowPage.php]: Anyway, rewriting it to `getAssignedIP` will allow us for some flexibility, we should avoid creating page methods that return booleans.

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners.xml [hunk 2]
```diff
@@ -86,5 +86,10 @@
             <argument>_sylius.cart</argument>
             <tag name="kernel.event_subscriber" />
         </service>
+        <service id="sylius.listener.order_customer_ip" class="Sylius\Bundle\CoreBundle\EventListener\OrderCustomerIpListener">
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners.xml]: Why aren't we doing it during order processing, just as with prices and everything else?

File: src/Sylius/Bundle/CoreBundle/Assigner/IpAssignerInterface.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Sylius\Bundle\CoreBundle\Assigner;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface IpAssignerInterface
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Assigner/IpAssignerInterface.php]: This interface seems quite useless to be honest, couldn't think of another implementation of IP assigner than the one provided.

File: src/Sylius/Bundle/CoreBundle/EventListener/OrderCustomerIpListener.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * @param GenericEvent $event
+     */
+    public function assignCustomerIpToOrder(GenericEvent $event)
+    {
+        $subject = $event->getSubject();
+        Assert::isInstanceOf($subject, OrderInterface::class);
+
+        $this->ipAssigner->assign($subject, $this->requestStack->getCurrentRequest());
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/EventListener/OrderCustomerIpListener.php]: We should try to assign customer IP to an order only on master request.

File: src/Sylius/Bundle/CoreBundle/Assigner/IpAssignerInterface.php (hunk 3 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Assigner/IpAssignerInterface.php]: Ip address send in an additional field of request. Can occur when one has a middleware between client and sylius instance

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners.xml [hunk 5]
```diff
@@ -86,5 +86,10 @@
...
+            <argument type="service" id="sylius.customer_ip_assigner" />
+            <argument type="service" id="request_stack" />
+            <tag name="kernel.event_listener" event="sylius.order.pre_summary" method="assignCustomerIpToOrder" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners.xml]: The event dispatched on order complete should be renamed from `summary` to `complete` (separate PR).

File: features/checkout/assigning_customer_ip_address_to_places_order.feature [hunk 6]
```diff
@@ -0,0 +1,22 @@
+@checkout
+Feature: Assigning customer's IP address to a placed order
+    In order to know from which IP address a new order has been places
```
Comment (lchrusciel) [inline features/checkout/assigning_customer_ip_address_to_places_order.feature]: placed

File: features/checkout/assigning_customer_ip_address_to_places_order.feature [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        And there is an administrator "sylius@example.com" identified by "sylius"
+        And there is a customer account "customer@example.com" identified by "sylius"
+        And I am logged in as "customer@example.com"
+
+    @ui
+    Scenario: Assigning customer's IP address to a newly placed order
+        Given I have product "PHP T-Shirt" in the cart
+        And I specified the shipping address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I proceed with "Free" shipping method and "Offline" payment
+        And I confirm my order
```
Comment (michalmarcinkowski) [inline features/checkout/assigning_customer_ip_address_to_places_order.feature]:
```
Given I added product "PHP T-Shirt" in the cart
And I specified the shipping ...
And I proceeded with ...
When I confirm my order
Then the administrator ...
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml [hunk 8]
```diff
@@ -79,6 +79,7 @@
         </many-to-many>
 
         <field type="string" name="tokenValue" column="token_value" nullable="true" />
+        <field type="string" name="customerIp" column="customer_ip" nullable="true" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml]: Should we add it to `Core` or `Order` bundle? /cc @pjedrzejewski

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml]: I've added it to `Core`, as I don't thing assigning customer IP is strictly connected to concept of `Order`s, it's more like extra functionality in *Sylius*. But of course /cc @pjedrzejewski, as I may be wrong 😄

**Comment by michalmarcinkowski** [comment]
Thanks Mateusz! Please correct Behats and rename the checkout summary event in a separate PR.

**Pull request merged**